lib/kafo/validator.rb

```diff
@@ -2,79 +2,111 @@
+        unless arg.to_s.start_with?('/')
+          @logger.error "#{arg.inspect} is not an absolute path."
+          return false
+        end
       end
+      return true
+    end
 
-      @params = params
-      @logger = KafoConfigure.logger
+    def validate_array(args)
+      args.each do |arg|
+        unless arg.is_a?(Array)
+          @logger.error "#{arg.inspect} is not a valid array."
+          return false
+        end
+      end
+      return true
+    end
 
-      @cache ||= Hash.new do |hash, key|
-        @logger.debug "Looked for #{key}"
-        param     = @params.select { |p| p.name == key.to_s }.first
-        hash[key] = param.nil? ? nil : param.value
+    def validate_bool(args)
+      args.each do |arg|
+        unless arg.is_a?(TrueClass) || arg.is_a?(FalseClass)
+          @logger.error "#{arg.inspect} is not a valid boolean."
+          return false
+        end
       end
+      return true
     end
 
-    def lookupvar(name, options = {})
-      @cache[name]
+    def validate_hash(args)
+      args.each do |arg|
+        unless arg.is_a?(Hash)
+          @logger.error "#{arg.inspect} is not a valid hash."
+          return false
+        end
+      end
+      return true
     end
 
-    # for puppet >= 3
-    def include?(value)
-      true
+    def validate_integer(args)
+      value = args[0]
+      max = args[1]
+      min = args[2]
+      int = Integer(value.to_s)
+      if min && int < min.to_i
+        @logger.error "#{value} must be at least #{min}."
+        return false
+      end
+      if max && int > max.to_i
+        @logger.error "#{value} must be less than #{max}."
+        return false
+      end
+      return true
+    rescue TypeError, ArgumentError
+      @logger.error "#{value.inspect} is not a valid integer."
+      return false
     end
 
-    # for puppet >= 3
-    def [](value, *args)
-      lookupvar(value)
+    def validate_listen_on(args)
+      valid_values = ['http', 'https', 'both']
+      args.each do |arg|
+        unless valid_values.include?(arg)
+          @logger.error "#{arg.inspect} is not a valid value.  Valid values are: #{valid_values.join(", ")}"
+          return false
+        end
+      end
+      return true
     end
 
-    def method_missing(method, *args, &block)
-      method.to_s =~ /^function_(.*)$/
-      super unless $1
-      super unless Puppet::Parser::Functions.function($1)
-      # In odd circumstances, this might not end up defined by the previous
-      # method, so we might as well be certain.
-      if engine.respond_to? method
-        @logger.debug "calling #{method.inspect} with #{args.inspect}"
-        engine.send(method, *args)
-        true
+    def validate_re(args)
+      value = args[0]
+      regexes = args[1]
+      regexes = [regexes] unless regexes.is_a?(Array)
+      message = args[2] || "#{value.inspect} does not match the accepted inputs: #{regexes.join(", ")}"
+
+      if regexes.all? { |rx| value =~ Regexp.compile(rx) }
```

Comment:
Good catch, I've updated the interpret_validation_args method to handle arrays (things that respond to #to_a) and recursively convert those arguments into regular objects.  I found my validate_re implementation was also incorrect (it checked for all?, not any?), so that's fixed too.

Lastly, I've fixed handling of the error message argument in interpret_validation_args as it was a "concat" AST object (string with interpolation), causing errors in the logger when the validation failed.